educator/src/Dscp/DB/SQLite/Util.hs

```diff
@@ -1,79 +1,213 @@
+    gPackPk = error "impossible"
+
+type PrimaryKeyWrapper pk inner = (Generic pk, GPrimaryKeyWrapper (G.Rep pk) inner)
+
+-- | Lift an entity to primary key, effectively just wrappes it into 'PrimaryKey' constructor.
+-- Works only for primary keys which consists of one item.
+packPk :: PrimaryKeyWrapper pk inner => inner -> pk
+packPk = G.to . gPackPk
+
+-- | Unwrap 'PrimaryKey' constructor.
+-- Works only for primary keys which consists of one item.
+unpackPk :: PrimaryKeyWrapper pk inner => pk -> inner
+unpackPk = gUnpackPk . G.from
+
+-- | Wrap an entiry into a primary key sql value.
+valPk_
+    :: (Beam.SqlValable pk, PrimaryKeyWrapper (Beam.HaskellLiteralForQExpr pk) inner)
+    => inner -> pk
+valPk_ = val_ . packPk
+
+-- | Quick way to fetch a single entiry refered by the given primary key.
+selectByPk
+    :: (MonadIO m, HasCallStack, _)
+    => (row Identity -> res)
+    -> Beam.DatabaseEntity be db (TableEntity table)
+    -> pk
+    -> DBT t w m (Maybe res)
+selectByPk mapper tbl key =
+    fmap fetchOne $
+    runSelect . Beam.select $
+    filter_ (valPk_ key `references_`) $
+    all_ tbl
+  where
+    fetchOne [] = Nothing
+    fetchOne l  = Just (mapper $ oneOrError l)
+
+-- | Quick way to check whether an entiry refered by the given primary key exists.
+existsWithPk
+    :: _
+    => Beam.DatabaseEntity be db (TableEntity table) -> pk -> DBT t w m Bool
+existsWithPk tbl key =
+    checkExists $ do
+        row <- all_ tbl
+        guard_ (Beam.pk row ==. valPk_ key)
+
+-- | Quick way to delete a single entiry refered by the given primary key.
+deleteByPk
+    :: (MonadIO m, _)
+    => Beam.DatabaseEntity be db (TableEntity table)
+    -> pk
+    -> DBT t 'Writing m ()
+deleteByPk tbl key = runDelete $ delete tbl (valPk_ key `references_`)
+
+-- | SQL CURRENT_TIMESTAMP function.
+currentTimestampUtc_
+    :: forall ctxt syntax s.
+       Beam.IsSql92ExpressionSyntax syntax
+    => Beam.QGenExpr ctxt syntax s UTCTime
+currentTimestampUtc_ =
+    -- The behavior should not depend much on the backend.
+    -- CURRENT_TIMESTAMP is nicely fetched both for "datetime" and "timestamp" types.
+    coerce (Beam.currentTimestamp_ @syntax)
+
+-- | Make a filtering predicate checking for values match.
+filterMatches_
+    :: _
+    => Maybe a -> QGenExpr syntax ctx s a -> QGenExpr syntax ctx s Bool
+filterMatches_ = \case
+    Nothing -> \_ -> val_ True
+    Just x  -> (==. val_ x)
+
+-- | Make a filtering predicate checking for values match.
+filterMatchesPk_
+    :: _
+    => Maybe a
+    -> PrimaryKey row (QGenExpr syntax ctx s)
+    -> QGenExpr syntax ctx s Bool
+filterMatchesPk_ = \case
+    Nothing -> \_ -> val_ True
+    Just x  -> (==. valPk_ x)
+
+-- | Make a filtering guard checking for values match.
+guardFilterMatches_
+    :: _
+    => Maybe a
+    -> QGenExpr syntax ctx s a
+    -> Beam.Q select db s ()
+guardFilterMatches_ = \case
+    Nothing -> \_ -> pass
+    Just x  -> guard_ . (==. val_ x)
```

Comment:
Might have just be `guardFilterMatches_ = guard_ . filterMatches_`, it seems (same for `guardFilterMatchesPk_`).